packages/mds-event-server/index.ts

```diff
@@ -1,18 +1,123 @@
 import express from 'express'
-import logger from '@mds-core/mds-logger'
+import uuid from 'uuid'
+import stan from 'node-nats-streaming'
 import { pathsFor } from '@mds-core/mds-utils'
+import log from '@mds-core/mds-logger'
 import { AboutRequestHandler, HealthRequestHandler, JsonBodyParserMiddleware } from '@mds-core/mds-api-server'
-import { Cloudevent, BinaryHTTPReceiver } from 'cloudevents-sdk/v1'
+import Cloudevent, { BinaryHTTPReceiver } from 'cloudevents-sdk/v1'
 
-export type EventProcessor<TData, TResult> = (type: string, data: TData, event: Cloudevent) => Promise<TResult>
+export type EventProcessor<TData, TResult> = (type: string, data: TData) => Promise<TResult>
+export type CEEventProcessor<TData, TResult> = (type: string, data: TData, event: Cloudevent) => Promise<TResult>
+
+const SUBSCRIPTION_TYPES = ['event', 'telemetry'] as const
+type SUBSCRIPTION_TYPE = typeof SUBSCRIPTION_TYPES[number]
+
+const subscriptionCb = async <TData, TResult>(processor: EventProcessor<TData, TResult>, msg: stan.Message) => {
+  const { TENANT_ID = 'mds' } = process.env
+
+  const TENANT_REGEXP = new RegExp(`^${TENANT_ID}\\.`)
+
+  try {
+    const {
+      spec: {
+        payload: { data, type }
+      }
+    } = JSON.parse(msg.getRawData().toString())
+
+    const parsedData = JSON.parse(data)
+
+    await processor(type.replace(TENANT_REGEXP, ''), parsedData)
+    msg.ack()
+  } catch (err) {
+    msg.ack()
+    await log.error(err)
+  }
+}
+
+const natsSubscriber = async <TData, TResult>({
+  nats,
+  processor,
+  TENANT_ID,
+  type
+}: {
+  nats: stan.Stan
+  processor: EventProcessor<TData, TResult>
+  TENANT_ID: string
+  type: SUBSCRIPTION_TYPE
+}) => {
+  const subscriber = nats.subscribe(`${TENANT_ID ?? 'mds'}.${type}`, {
+    ...nats.subscriptionOptions(),
+    manualAcks: true,
+    maxInFlight: 1
+  })
+
+  subscriber.on('message', async (msg: stan.Message) => {
+    return subscriptionCb(processor, msg)
+  })
+}
+
+const initializeNatsClient = ({
+  NATS,
+  STAN_CLUSTER,
+  STAN_CREDS
+}: {
+  NATS: string
+  STAN_CLUSTER: string
+  STAN_CREDS?: string
+}) => {
+  return stan.connect(STAN_CLUSTER, `mds-event-processor-${uuid()}`, {
+    url: `nats://${NATS}:4222`,
```

Comment:
Should this `url` also be constructed from an environment variable? ... using `NATS_PORT` perhaps?

Reply:
+1 let's put the entire url in helm ... more declaratively obvious, where other like resources reside, etc.